File: server/routes/game.ts
import { Router } from 'express';
import {
  getGameSession,
  getReverseFieldsForSession,
  getReverseValuesForSession,
  nextQuestion,
  quitGame,
  startGame,
  submitGuess,
  submitReverseQuery,
} from '../services/gameService';
import { searchCharacters } from '../services/dataLoader';
import { DailyAlreadyPlayedError, GameMode, ReverseCondition, Theme } from '../types';
import { getClientKey } from '../utils/clientKey';

export const gameRouter = Router();

const VALID_THEMES: Theme[] = ['csgo', 'football', 'nba', 'anime', 'pokemon'];
const VALID_MODES: GameMode[] = [
  'classic-six',
  'daily-one',
  'progressive-hint',
  'reverse-bomb',
];

gameRouter.post('/start', (req, res) => {
  try {
    const { playerName, theme, gameMode } = req.body as {
      playerName?: string;
      theme?: Theme;
      gameMode?: GameMode;
    };
    if (!playerName?.trim()) {
      return res.status(400).json({ error: '请输入昵称' });
    }
    if (!theme || !VALID_THEMES.includes(theme)) {
      return res.status(400).json({ error: '请选择有效主题' });
    }
    const mode =
      gameMode && VALID_MODES.includes(gameMode) ? gameMode : 'classic-six';
    const session = startGame(playerName, theme, mode, getClientKey(req));
    res.json(session);
  } catch (e) {
    if (e instanceof DailyAlreadyPlayedError) {
      return res.status(409).json({
        error: e.message,
        code: e.code,
        sessionId: e.sessionId,
      });
    }
    res.status(500).json({ error: (e as Error).message });
  }
});

gameRouter.get('/suggest', (req, res) => {
  try {
    const theme = req.query.theme as Theme;
    const q = String(req.query.q || '');
    if (!theme || !VALID_THEMES.includes(theme)) {
      return res.status(400).json({ error: '无效主题' });
    }
    const results = searchCharacters(theme, q, 8);
    res.json({ results });
  } catch (e) {
    res.status(500).json({ error: (e as Error).message });
  }
});

gameRouter.post('/guess', (req, res) => {
  try {
    const { sessionId, guessText, characterId } = req.body as {
      sessionId?: string;
      guessText?: string;
      characterId?: string;
    };
    if (!sessionId || !guessText?.trim()) {
      return res.status(400).json({ error: '缺少参数' });
    }
    const result = submitGuess(sessionId, guessText, characterId);
    res.json(result);
  } catch (e) {
    res.status(500).json({ error: (e as Error).message });
  }
});

gameRouter.post('/next', (req, res) => {
  try {
    const { sessionId } = req.body as { sessionId?: string };
    if (!sessionId) return res.status(400).json({ error: '缺少 sessionId' });
    const session = nextQuestion(sessionId);
    res.json(session);
  } catch (e) {
    res.status(500).json({ error: (e as Error).message });
  }
});

gameRouter.post('/quit', (req, res) => {
  try {
    const { sessionId } = req.body as { sessionId?: string };
    if (!sessionId) return res.status(400).json({ error: '缺少 sessionId' });
    const session = quitGame(sessionId);
    res.json(session);
  } catch (e) {
    res.status(500).json({ error: (e as Error).message });
  }
});

gameRouter.get('/reverse-fields', (req, res) => {
  try {
    const sessionId = String(req.query.sessionId || '');
    if (!sessionId) return res.status(400).json({ error: '缺少 sessionId' });
    const fields = getReverseFieldsForSession(sessionId);
    res.json({ fields });
  } catch (e) {
    const msg = (e as Error).message;
    if (msg === 'Session not found') return res.status(404).json({ error: msg });
    res.status(400).json({ error: msg });
  }
});

gameRouter.get('/reverse-values', (req, res) => {
  try {
    const sessionId = String(req.query.sessionId || '');
    const field = String(req.query.field || '');
    if (!sessionId || !field) {
      return res.status(400).json({ error: '缺少 sessionId 或 field' });
    }
    const values = getReverseValuesForSession(sessionId, field);
    res.json(values);
  } catch (e) {
    const msg = (e as Error).message;
    if (msg === 'Session not found') return res.status(404).json({ error: msg });
    res.status(400).json({ error: msg });
  }
});

gameRouter.post('/reverse-query', (req, res) => {
  try {
    const { sessionId, condition } = req.body as {
      sessionId?: string;
      condition?: ReverseCondition;
    };
    if (!sessionId || !condition?.field || !condition?.operator) {
      return res.status(400).json({ error: '缺少参数' });
    }
    const result = submitReverseQuery(sessionId, condition);
    res.json(result);
  } catch (e) {
    const msg = (e as Error).message;
    if (msg === 'Session not found') return res.status(404).json({ error: msg });
    if (msg === '提问机会已用完') return res.status(409).json({ error: msg });
    res.status(400).json({ error: msg });
  }
});

gameRouter.get('/:sessionId', (req, res) => {
  try {
    const session = getGameSession(req.params.sessionId);
    if (!session) return res.status(404).json({ error: '会话不存在' });
    res.json(session);
  } catch (e) {
    res.status(500).json({ error: (e as Error).message });
  }
});

File: server-go/internal/api/game.go
package api

import (
	"encoding/json"
	"errors"
	"net/http"
	"strconv"
	"strings"

	"guess-who/server-go/internal/services"
	"guess-who/server-go/internal/types"

	"github.com/go-chi/chi/v5"
)

var validThemes = map[types.Theme]bool{
	types.ThemeCSGO: true, types.ThemeFootball: true, types.ThemeNBA: true, types.ThemePokemon: true,
}

var validModes = map[types.GameMode]bool{
	types.ModeClassicSix: true, types.ModeDailyOne: true,
	types.ModeProgressive: true, types.ModeReverseBomb: true,
}

// RegisterGameRoutes 挂载游戏 REST API 端点。
func RegisterGameRoutes(r chi.Router) {
	r.Post("/api/game/start", handleStart)
	r.Get("/api/game/suggest", handleSuggest)
	r.Post("/api/game/guess", handleGuess)
	r.Post("/api/game/next", handleNext)
	r.Post("/api/game/quit", handleQuit)
	r.Get("/api/game/reverse-fields", handleReverseFields)
	r.Get("/api/game/reverse-values", handleReverseValues)
	r.Post("/api/game/reverse-query", handleReverseQuery)
	r.Get("/api/game/{sessionId}", handleGetSession)
	r.Get("/api/leaderboard", handleLeaderboard)
	r.Get("/api/leaderboard/daily/today", handleDailyToday)
	r.Get("/api/health", handleHealth)
	r.Get("/api/ladder/invites", handleLadderInvites)
}

func writeJSON(w http.ResponseWriter, status int, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func handleStart(w http.ResponseWriter, r *http.Request) {
	var req struct {
		PlayerName string         `json:"playerName"`
		Theme      types.Theme    `json:"theme"`
		GameMode   types.GameMode `json:"gameMode"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request")
		return
	}

	playerKey := services.GetClientKey(r)
	sess, err := services.StartSession(req.PlayerName, req.Theme, req.GameMode, playerKey)
	if err != nil {
		var dailyErr *types.DailyAlreadyPlayedError
		if errors.As(err, &dailyErr) {
			writeJSON(w, http.StatusConflict, map[string]interface{}{
				"error":     dailyErr.Error(),
				"code":      dailyErr.Code,
				"sessionId": dailyErr.SessionID,
			})
			return
		}
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, sess)
}

func handleSuggest(w http.ResponseWriter, r *http.Request) {
	theme := types.Theme(r.URL.Query().Get("theme"))
	q := strings.TrimSpace(r.URL.Query().Get("q"))
	results := services.SuggestCharacters(theme, q)
	writeJSON(w, http.StatusOK, map[string]interface{}{"results": results})
}

func handleGuess(w http.ResponseWriter, r *http.Request) {
	var req struct {
		SessionID   string  `json:"sessionId"`
		GuessText   string  `json:"guessText"`
		CharacterID *string `json:"characterId"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request")
		return
	}

	resp, err := services.SubmitGuess(req.SessionID, req.GuessText, req.CharacterID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

func handleNext(w http.ResponseWriter, r *http.Request) {
	var req struct {
		SessionID string `json:"sessionId"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request")
		return
	}
	sess, err := services.NextQuestion(req.SessionID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, sess)
}

func handleQuit(w http.ResponseWriter, r *http.Request) {
	var req struct {
		SessionID string `json:"sessionId"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request")
		return
	}
	sess, err := services.QuitGame(req.SessionID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, sess)
}

func handleReverseFields(w http.ResponseWriter, r *http.Request) {
	sessionID := r.URL.Query().Get("sessionId")
	fields, err := services.GetReverseFieldsForSession(sessionID)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, fields)
}

func handleReverseValues(w http.ResponseWriter, r *http.Request) {
	sessionID := r.URL.Query().Get("sessionId")
	field := r.URL.Query().Get("field")
	resp, err := services.GetReverseValuesForSession(sessionID, field)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

func handleReverseQuery(w http.ResponseWriter, r *http.Request) {
	var req struct {
		SessionID string                 `json:"sessionId"`
		Condition types.ReverseCondition `json:"condition"`
	}
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "bad request")
		return
	}
	resp, err := services.SubmitReverseQuerySession(req.SessionID, req.Condition)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, resp)
}

func handleGetSession(w http.ResponseWriter, r *http.Request) {
	sessionID := chi.URLParam(r, "sessionId")
	sess, err := services.GetSession(sessionID)
	if err != nil {
		writeError(w, http.StatusNotFound, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, sess)
}

func handleLeaderboard(w http.ResponseWriter, r *http.Request) {
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	if limit <= 0 {
		limit = 20
	}
	gameMode := types.GameMode(r.URL.Query().Get("gameMode"))
	if gameMode == "" {
		gameMode = types.ModeClassicSix
	}
	if !validModes[gameMode] {
		writeError(w, http.StatusBadRequest, "无效模式")
		return
	}

	themeStr := r.URL.Query().Get("theme")
	var themePtr *types.Theme
	if themeStr != "" {
		theme := types.Theme(themeStr)
		if !validThemes[theme] {
			writeError(w, http.StatusBadRequest, "无效主题")
			return
		}
		themePtr = &theme
	}

	if gameMode == types.ModeDailyOne {
		if themePtr == nil {
			writeError(w, http.StatusBadRequest, "每日榜需指定主题")
			return
		}
		entries, err := services.GetDailyLeaderboard(*themePtr, limit, "")
		if err != nil {
			writeError(w, http.StatusInternalServerError, err.Error())
			return
		}
		writeJSON(w, http.StatusOK, entries)
		return
	}

	entries, err := services.GetLeaderboard(limit, gameMode, themePtr)
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, entries)
}

func handleDailyToday(w http.ResponseWriter, r *http.Request) {
	theme := types.Theme(r.URL.Query().Get("theme"))
	if !validThemes[theme] {
		writeError(w, http.StatusBadRequest, "无效主题")
		return
	}
	info, err := services.GetDailyTodayInfo(theme, services.GetClientKey(r))
	if err != nil {
		writeError(w, http.StatusInternalServerError, err.Error())
		return
	}
	writeJSON(w, http.StatusOK, info)
}

func handleHealth(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]string{"status": "ok", "backend": "go"})
}

func handleLadderInvites(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, services.ListLadderInvites())
}
